alphagov/paas-nats-release
# WIP: [#124720541] Implement logic to send service logs to syslog

# What

Several of the CF and Diego platform services send logs to the local syslog, tagged as `vcap.*`. . The [metron agent from loggregrator configures the local syslog to send the logs to metron](https://github.com/cloudfoundry/loggregator/blob/develop/jobs/metron_agent/spec#L29) so that later can be sent to a centralised logging service. 


In some cases it is the service application itself who sends the logs to syslog, but it seems to be a common pattern across CF and Diego services to [let the service log to stdout/stderr, and redirect that output to local files or syslog](https://github.com/cloudfoundry/diego-release/blob/develop/jobs/bbs/templates/bbs_as_vcap.erb#L91) using the [`logger` command](http://linux.die.net/man/1/logger). 

Not all the services implement this feature, and instead they only log to local files. In order to centralise all the logs, so we will add the logic required to send the logs of the missing services. 

We will implement the *log to stdout and redirect  to `logger` pattern* whenever is possible.

## Implementation

We add a property `nats.log_to_syslog`. When true, we redirect the stdout and stderr in the ctl script into syslog following the standard vcap pattern, in addition to files.

[Nats will print the logs to stdout if no log option is specified](http://nats.io/documentation/server/gnatsd-logging/), so enabling this will remove the `-l` option passed to nats.

## How to review? 

Currently you can test it manually by:

 1. Download the `cf-manifest.yml` and add the property `nats.log_to_syslog: true`
 2. Render the ctl script using our [bosh render.rb tool from paas-cf](https://github.com/alphagov/paas-cf/blob/master/platform-tests/bosh-template-renderer/render.rb): `./platform-tests/bosh-template-renderer/render.rb ~/workspace/paas-nats-release/jobs/nats/templates/nats_ctl.erb  ~/workspace/paas-nats-release/jobs/nats/spec /tmp/cf-manifest.yml`
 3. Copy this script to one of the nats VMs running in a deployed environment, in `/var/vcap/jobs/nats/bin/nats_ctl`
 4. Restart nats and check if the logs are being sent to kibana. 

## Who? 

Anyone but @richardc or @keymon
